shell/gpu/gpu_surface_metal.mm

```diff
@@ -112,11 +139,25 @@ GrBackendRenderTarget metal_render_target(bounds.width * scale,   // width
     return nullptr;
   }
 
-  auto submit_callback = [drawable = next_drawable, command_buffer](
+  // External views need to present with transaction. When presenting with
+  // transaction, we have to block, otherwise we risk presenting the drawable
+  // after the CATransaction has completed.
+  // See:
+  // https://developer.apple.com/documentation/quartzcore/cametallayer/1478157-presentswithtransaction
+  // TODO(dnfield): only do this if transactions are actually being used.
+  // https://github.com/flutter/flutter/issues/24133
+  bool wait = delegate_->GetExternalViewEmbedder() != nullptr;
+
+  auto submit_callback = [drawable = next_drawable, command_buffer, wait](
                              const SurfaceFrame& surface_frame, SkCanvas* canvas) -> bool {
     canvas->flush();
-    [command_buffer.get() presentDrawable:drawable.get()];
     [command_buffer.get() commit];
+    if (!wait) {
```

Comment:
IIRC, `presentDrawable` is just waiting for the command buffer to be scheduled and the and then presenting the same. Since you want the wait to happen now, the following code should be equivalent and more readable.

```
if (wait) {
 [command_buffer waitUntilScheduled];
}
[command_buffer presentDrawable:drawable];
```

Reply:
I'm a little concerned that `presentDrawable` could end up unexpectedly yielding - the documentation on `presentsWithTransaction` says to use `present` instead of `presentDrawable` for this reason.

Reply:
Just had to update this - the current formulation doesn't work, we have to call the `presentDrawable` before we commit otherwise it throws an exception sometimes.